raft/server.cc

```diff
@@ -1292,7 +1305,21 @@ future<> server_impl::set_configuration(config_member_set c_new, seastar::abort_
     // would be the one corresponding to our joint configuration,
     // no matter if the leader changed in the meantime.
 
-    auto f = _non_joint_conf_commit_promise.emplace().get_future();
+    auto guard = defer([this] {
+        // If everything goes right, `io_fiber` will reset the field,
+        // but in case of failure - ensure it's reset here.
+        _non_joint_conf_commit_promise.reset();
```

Comment:
What failure? How can we destroy the promise without signalling it after taking the future below?

Reply:
The `wait_for_entry` call may throw, for example

Reply:
>  If we missed resetting optional there just reset it. 

That's what the guard does.

Reply:
> The guard assumes that we can abandon the promise at any point without any addition action.

The guard doesn't assume anything, it only ensures that the promise is reset in all cases (it doesn't mean that some cases won't need additional action).

Reply:
So why split error handling into two different parts. The code just after the guard takes `f`. From that point on the guard cannot be the only thing that handles an error. Put the reset in the error handling that  handles `f` already. It is there.

Reply:
Ok